pycalphad/core/eqsolver.pyx

```diff
@@ -84,41 +82,43 @@ cdef bint add_new_phases(object composition_sets, object removed_compsets, objec
         return True
     return False
 
-# composition_sets: List[CompositionSet]
-# comps: List[v.Species]
-# cur_conds: OrderedDict[str, float]
-# iter_solver: SolverBase instance
-cpdef pointsolve(composition_sets, comps, cur_conds, iter_solver):
-    "Mutates composititon_sets with updated values if it converges. Returns SolverResult."
-    return _solve_and_update_if_converged(composition_sets, comps, cur_conds, Problem, iter_solver)
 
-cdef _solve_and_update_if_converged(composition_sets, comps, cur_conds, problem, iter_solver):
-    "Mutates composititon_sets with updated values if it converges. Returns SolverResult."
+cpdef update_composition_sets(composition_sets, solver_result, remove=True):
     cdef CompositionSet compset
-    prob = problem(composition_sets, comps, cur_conds)
-    result = iter_solver.solve(prob)
-    composition_sets = prob.composition_sets
-    x = result.x
+    x = solver_result.x
     compset = composition_sets[0]
-    var_offset = len(compset.phase_record.state_variables)
+    num_compsets = len(composition_sets)
+    num_state_variables = len(compset.phase_record.state_variables)
+    var_offset = num_state_variables
+    num_vars = sum([compset.phase_record.phase_dof for compset in composition_sets]) + num_compsets + num_state_variables
     phase_idx = 0
     compsets_to_remove = []
     for compset in composition_sets:
-        phase_amt = x[prob.num_vars - prob.num_phases + phase_idx]
+        phase_amt = x[num_vars - num_compsets + phase_idx]
         # Mark unstable phases for removal
         if phase_amt == 0.0 and not compset.fixed:
             compsets_to_remove.append(int(phase_idx))
         compset.update(x[var_offset:var_offset + compset.phase_record.phase_dof],
-                       phase_amt, x[:len(compset.phase_record.state_variables)])
+                       phase_amt, x[:num_state_variables])
         var_offset += compset.phase_record.phase_dof
         phase_idx += 1
-    # Watch removal order here, as the indices of composition_sets are changing!
-    for idx in reversed(compsets_to_remove):
-        del composition_sets[idx]
+    if remove:
+        # Watch removal order here, as the indices of composition_sets are changing!
+        for idx in reversed(compsets_to_remove):
+            del composition_sets[idx]
+
+
+# composition_sets: List[CompositionSet]
+# cur_conds: OrderedDict[str, float]
```

Comment:
What do you think about allowing a `OrderedDict[v.StateVariable, float]` to be passed in here, and taking care of the conditions dict conversion for the user? The `solve` interface will remain available for users who don't want to pay the performance penalty.

Reply:
I thought about this some more, and it's okay to defer a decision for now, since we can always add support for it later. I anticipate more refactoring of this part of the core (if not this exact function) in the near future, so we may want to see how our revised handling of, e.g., starting points interacts with this function.